ruby_programming/advanced_ruby/blocks.md

```diff
@@ -0,0 +1,547 @@
+
+Where might this be useful? Well think about each time you use `each`. You need to yield each member of the collection to the block. Say you're writing a banking api, it will allow other banks to use your library to handle all the banking matters. One such api is a method that prints all of the transactions to a statement. For our example we'll use the terminal, but it could be a spreadsheet or anything. One issue is that different banks may want to print the transaction with different formats. We can iterate through the transactions and for each one we can yield it to a block that the caller of your method attaches to the method call. They can define how the transactions will be printed to their statement and you can focus on delivering bug free banking transactions
+
+~~~ruby
+@transactions = [10, -15, 25, 30, -24, -70, 999]
+
+def transaction_statement
+  @transactions.each do |transaction|
+    yield transaction # You just yield the transaction amount
+  end
+end
+
+transaction_statement do |transaction|
+  puts "%0.2f" % transaction # The person who calls the method can define how it is handled.
+end
+~~~
+
+If another bank wanted to print their transactions another way it's no problem, they can supply their own block.
+
+What if instead you didn't want the caller to define where the transaction for output, but just the format? Like any method call in Ruby there is a return value. So when you call yield, it returns the last executed value from the block. Using the example above if you instead moved the `puts` from the block to inside the each loop your method would be in control in where the transaction were printed. Maybe you only allow exporting to a csv for example.
+
+~~~ruby
+@transactions = [10, -15, 25, 30, -24, -70, 999]
+
+def transaction_statement
+  @transactions.each do |transaction|
+    puts yield transaction
+  end
+end
+
+transaction_statement do |transaction|
+  "%0.2f" % transaction
+end
+~~~
+
+If you want to gather the value returned from the block, you can just assign it to a variable or collect it in a data structure.
+
+~~~ruby
+@transactions = [10, -15, 25, 30, -24, -70, 999]
+
+def transaction_statement
+  formatted_transactions = []
+  @transactions.each do |transaction|
+    formatted_transactions << yield(transaction) # I've put () around transaction just for clarity here but they aren't required.
+  end
+  
+  formatted_transactions
+end
+
+transaction_statement do |transaction|
+  "%0.2f" % transaction
+end
+~~~
+
+You can also write explicit return statements from a block and it will return whatever value is after that. This works the same way as an explicit return from a method. This might be useful if you need some kind of guard clause.
```

Comment:
Do you think it would be worth wrapping "guard clause" in a link to explain what that is? just incase the student hasn't come across that term before.

Reply:
Yeah. I found a decent link on Ruby because the Wiki page on guard clauses is a bit too technical I think. Let me know what you think.